File: CHAPTER 4. 옵셔널에서 널이 될 수 있는 타입으로/yckim/chapter4/java/v1/Legs.java
package chapter4.java.v1;

import chapter4.kotlin.Leg;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

public class Legs {

    public static Optional<Leg> findLongestLegOver(
            List<Leg> legs,
            Duration duration
    ) {
        Leg result = null;
        for (Leg leg : legs) {
            if (result == null || isLongerThan(leg, result.getPlannedDuration())) {
                result = leg;
            }
        }
        return Optional.ofNullable(result);
    }

    private static boolean isLongerThan(Leg leg, Duration duration) {
        return leg.getPlannedDuration().compareTo(duration) > 0;
    }
}

File: CHAPTER 04. 옵셔널에서 널이 될 수 있는 타입으로/yckim/chapter4/kotlin/v3/Legs.kt
package chapter4.kotlin.v3

import chapter4.kotlin.Leg
import java.time.Duration

fun List<Leg>.longestOver(
    legs: List<Leg>,
    duration: Duration
): Leg? =
    legs.maxByOrNull(Leg::plannedDuration)?.takeIf { longestLeg ->
        longestLeg.plannedDuration > duration
    }
